beater/agent_config_handler_integration_test.go

```diff
@@ -0,0 +1,110 @@
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package beater
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/elastic/beats/libbeat/common"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/apm-server/beater/beatertest"
+	"github.com/elastic/apm-server/beater/headers"
+	"github.com/elastic/apm-server/beater/request"
+	"github.com/elastic/apm-server/tests"
+)
+
+func TestAgentConfigHandler_RequireAuthorizationMiddleware(t *testing.T) {
+	t.Run("Unauthorized", func(t *testing.T) {
+		cfg := cfgEnabledACM()
+		cfg.SecretToken = "1234"
+		rec := requestToACMHandler(t, cfg)
+
+		assert.Equal(t, http.StatusUnauthorized, rec.Code)
+		tests.AssertApproveResult(t, acmApprovalPath(t.Name()), rec.Body.Bytes())
+	})
+
+	t.Run("Authorized", func(t *testing.T) {
+		cfg := cfgEnabledACM()
+		cfg.SecretToken = "1234"
+		h, err := agentConfigHandler(cfg, beatertest.NilReporter)
+		require.NoError(t, err)
+		c, rec := beatertest.DefaultContextWithResponseRecorder()
+		c.Request.Header.Set(headers.Authorization, "Bearer 1234")
+		h(c)
+
+		assert.Equal(t, http.StatusServiceUnavailable, rec.Code)
+		tests.AssertApproveResult(t, acmApprovalPath(t.Name()), rec.Body.Bytes())
+	})
+}
+
+func TestAgentConfigHandler_KillSwitchMiddleware(t *testing.T) {
+	t.Run("Off", func(t *testing.T) {
+		rec := requestToACMHandler(t, DefaultConfig(beatertest.MockBeatVersion()))
+
+		assert.Equal(t, http.StatusForbidden, rec.Code)
+		tests.AssertApproveResult(t, acmApprovalPath(t.Name()), rec.Body.Bytes())
+
+	})
+
+	t.Run("On", func(t *testing.T) {
+		rec := requestToACMHandler(t, cfgEnabledACM())
+
+		assert.Equal(t, http.StatusServiceUnavailable, rec.Code)
+		tests.AssertApproveResult(t, acmApprovalPath(t.Name()), rec.Body.Bytes())
+	})
+}
+
+func TestAgentConfigHandler_PanicMiddleware(t *testing.T) {
+	h, err := agentConfigHandler(DefaultConfig(beatertest.MockBeatVersion()), beatertest.NilReporter)
+	require.NoError(t, err)
+	rec := &beatertest.WriterPanicOnce{}
+	c := &request.Context{}
+	c.Reset(rec, httptest.NewRequest(http.MethodGet, "/", nil))
+	h(c)
+	assert.Equal(t, http.StatusInternalServerError, rec.StatusCode)
+	tests.AssertApproveResult(t, acmApprovalPath(t.Name()), rec.Body.Bytes())
+}
+
+func TestAgentConfigHandler_MonitoringMiddleware(t *testing.T) {
+	beatertest.ClearRegistry(serverMetrics, ACMResultIDToMonitoringInt)
```

Comment:
Having to call `ClearRegistry` before a test is not great; this calls a bit for a `TestMain` with a setup function that clears the registry before each unit test, and then having each unit test needing a clean registry in that file.

Reply:
I prefer keeping the integration test files logically together, but I change the monitoring testing method accordingly, so that `clear` does not need to be called from outside.